Updates for TensorKit compatibility

kshyatt opened the pull request:
The `ReshapedArray` overrides are needed to dispatch to the correct GPU algorithms. Needed to modify the type signature for the default algorithms to avoid ambiguities. Also it's nice to give some more info about dimension mismatches.

Jutho commented on src/implementations/lq.jl:
```diff
@@ -181,7 +181,7 @@ function lq_via_qr!(A::AbstractMatrix, L::AbstractMatrix, Q::AbstractMatrix,
     m, n = size(A)
     minmn = min(m, n)
-    At = adjoint!(similar(A'), A)::AbstractMatrix
+    At = min(m, n) > 0 ? adjoint!(similar(A'), A)::AbstractMatrix : similar(A')
```
Is this `::AbstractMatrix` type assert useful?

kshyatt commented:
Changed the format of some of the `orthnull` tests -- `CuArray` and `UniformScaling` and `isapprox` *really* don't play nicely together. Also added a path to use `CUSOLVER_QRIteration` even in the case where `m < n` via transpose. Happy for feedback on the implementation - should this be a separate algorithm?

Jutho commented on src/implementations/svd.jl:
```diff
@@ -288,7 +288,7 @@ function MatrixAlgebraKit.svd_compact!(A::AbstractMatrix, USVᴴ, alg::GPU_SVDAl
     if alg isa GPU_QRIteration
         isempty(alg.kwargs) ||
-            throw(ArgumentError("GPU_QRIteration does not accept any keyword arguments"))
+            @warn "GPU_QRIteration does not accept any keyword arguments"
```
I would need to check, but some of the LAPACK algorithm names are used for different factorisations, and might support keywords for one and not for the other, or might support a different set of keywords. We can also remove these explicit checks, just pass on `alg.kwargs` to the underlying LAPACK/CUSOLVER/rocSOLVER routine, and then have these complain if the set of keyword arguments is invalid.

Jutho commented on src/common/matrixproperties.jl:
```diff
@@ -42,6 +42,7 @@ See also [`isisometry`](@ref) and [`is_right_isometry`](@ref).
 """ is_left_isometry
 
 function is_left_isometry(A::AbstractMatrix; isapprox_kwargs...)
+    iszero(min(size(A)...)) && return true
```
I would say: `iszero(size(A, 2)) && return true` for `is_left_isometry` and `iszero(size(A,1))` for `is_right_isometry below`.

Jutho commented on src/implementations/svd.jl:
```diff
@@ -245,17 +245,43 @@ _gpu_gesvd!(A::AbstractMatrix, S::AbstractVector, U::AbstractMatrix, Vᴴ::Abstr
+    minmn = min(m, n)
+    At = min(m, n) > 0 ? adjoint!(similar(A'), A)::AbstractMatrix : similar(A')
+    Ut = similar(U')
+    Vᴴt = similar(Vᴴ')
```
Can we call these variables `Aᴴ`, `Uᴴ` and simple `V` ?

Jutho commented on src/implementations/svd.jl:
```diff
@@ -245,17 +245,43 @@ _gpu_gesvd!(A::AbstractMatrix, S::AbstractVector, U::AbstractMatrix, Vᴴ::Abstr
+    else
+        _gpu_gesvd!(At, S, Vᴴt, Ut)
+    end
+    length(U) > 0 ? adjoint!(U, Ut) : one!(U)
```
Why is the `one!(U)` necessary if `U` is empty, i.e. why not

```julia
length(U) > 0 && adjoint!(U, Ut)
```

kshyatt commented on src/implementations/svd.jl:
```diff
@@ -245,17 +245,43 @@ _gpu_gesvd!(A::AbstractMatrix, S::AbstractVector, U::AbstractMatrix, Vᴴ::Abstr
+    end
+    length(U) > 0 ? adjoint!(U, Ut) : one!(U)
+    length(Vᴴ) > 0 ? adjoint!(Vᴴ, Vᴴt) : one!(Vᴴ)
+    conj!(S)
```
Yes... the risks of making updates at 9pm at night

kshyatt commented on src/implementations/svd.jl:
```diff
@@ -245,17 +245,43 @@ _gpu_gesvd!(A::AbstractMatrix, S::AbstractVector, U::AbstractMatrix, Vᴴ::Abstr
+    Ut = similar(U')
+    Vᴴt = similar(Vᴴ')
+    if size(U) == (m, m)
+        _gpu_gesvd!(At, view(S, 1:minmn, 1), Vᴴt, Ut)
```
Here I'm following what the CPU bindings have done, but I suppose we could be reusing an S over and over between differently sized arrays?

lkdvos commented on ext/MatrixAlgebraKitCUDAExt/MatrixAlgebraKitCUDAExt.jl:
```diff
@@ -8,30 +8,49 @@ using MatrixAlgebraKit: LQViaTransposedQR, TruncationByValue, AbstractAlgorithm
 using LinearAlgebra: BlasFloat
 
+using CUDA: i32
+
```
```suggestion
```

I think this is already in the import list above
